internal/controller/templatemanagement_controller.go

```diff
@@ -48,9 +54,198 @@ func (r *TemplateManagementReconciler) Reconcile(ctx context.Context, req ctrl.R
 		l.Error(err, "Failed to get TemplateManagement")
 		return ctrl.Result{}, err
 	}
+
+	defer func() {
+		statusErr := ""
+		if err != nil {
+			statusErr = err.Error()
+		}
+		templateMgmt.Status.Error = statusErr
+		err = errors.Join(err, r.updateStatus(ctx, templateMgmt))
+	}()
+
+	currentState, err := templateutil.GetCurrentTemplatesState(ctx, r.Client, r.SystemNamespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	expectedState, targetNamespaces, err := templateutil.ParseAccessRules(ctx, r.Client, templateMgmt.Spec.AccessRules, currentState)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.ensureDefaultHelmRepository(ctx, targetNamespaces)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var errs error
+	err = r.distributeTemplates(ctx, expectedState.ClusterTemplatesState, templateutil.ClusterTemplateKind)
+	if err != nil {
+		errs = errors.Join(errs, err)
+	}
+	err = r.distributeTemplates(ctx, expectedState.ServiceTemplatesState, templateutil.ServiceTemplateKind)
+	if err != nil {
+		errs = errors.Join(errs, err)
+	}
+	if errs != nil {
+		return ctrl.Result{}, errs
+	}
+
+	templateMgmt.Status.ObservedGeneration = templateMgmt.Generation
+	templateMgmt.Status.Current = templateMgmt.Spec.AccessRules
 	return ctrl.Result{}, nil
 }
 
+func (r *TemplateManagementReconciler) updateStatus(ctx context.Context, templateMgmt *hmc.TemplateManagement) error {
+	if err := r.Status().Update(ctx, templateMgmt); err != nil {
+		return fmt.Errorf("failed to update status for TemplateManagement %s: %w", templateMgmt.Name, err)
+	}
+	return nil
+}
+
+func (r *TemplateManagementReconciler) ensureDefaultHelmRepository(ctx context.Context, namespaces map[string]bool) error {
+	l := log.FromContext(ctx)
+
+	source := &sourcev1.HelmRepository{}
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: r.SystemNamespace,
+		Name:      defaultRepoName,
+	}, source)
+	if err != nil {
+		return err
+	}
+	var errs error
+	for ns := range namespaces {
+		helmRepo := &sourcev1.HelmRepository{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: ns,
+				Name:      defaultRepoName,
+				Labels: map[string]string{
+					hmc.HMCManagedLabelKey: hmc.HMCManagedLabelValue,
+				},
+			},
+		}
+		err := r.Get(ctx, types.NamespacedName{
```

Comment:
for helmrepository in contrary to templates we can use createourupdate :D but let's decide if we want to distribute these resources or keep them all in the system namespace first